### ISSUE TITLE
solve_linear_system contains duplicate rref algorithm

### ISSUE THREAD

**Issue by asmeurer** [sympy/sympy#4949]
```
It seems that we have two different rref (reduced row echelon form) algorithms, one in 
Matrix.rref and one in solve_linear_system in solvers.py.  I don't know too much about these 
algorithms, so I don't know if they really are the same algorithm, but in either case, I think 
solve should just be calling Matrix.rref.

This might speed things up too (examples taken from Lay's Linear Algebra, Third Edition):

In [15]: c = Matrix([[1, -2, 1, 0],[0, 2, -8, 8],[-4, 5, 9, -9]])

In [16]: c
Out[16]: 
⎡1   -2  1   0 ⎤
⎢              ⎥
⎢0   2   -8  8 ⎥
⎢              ⎥
⎣-4  5   9   -9⎦

In [17]: c.rref()
Out[17]: 
⎛⎡1  0  0  29⎤, [0, 1, 2]⎞
⎜⎢           ⎥           ⎟
⎜⎢0  1  0  16⎥           ⎟
⎜⎢           ⎥           ⎟
⎝⎣0  0  1  3 ⎦           ⎠

In [19]: solve_linear_system(c, x, y, z)
Out[19]: {x: 29, y: 16, z: 3}

In [25]: %timeit solve_linear_system(c, x, y, z)
100 loops, best of 3: 11.2 ms per loop

In [26]: %timeit c.rref()
1000 loops, best of 3: 3.59 ms per loop

In [30]: e = Matrix([[4, -1, -1, 0, 0, 0, 0, 0, 5],[-1, 4, 0, -1, 0, 0, 0, 0, 15],[-1, 0, 4, -1, -1, 0, 0, 
0, 0],[0, -1, -1, 4, 0, -1, 0, 0, 10],[0, 0, -1, 0, 4, -1, -1, 0, 0],[0, 0, 0, -1, -1, 4, 0, -1, 10],[0, 0, 
0, 0, -1, 0, 4, -1, 20],[0, 0, 0, 0, 0, -1, -1, 4, 30]])

In [32]: e.rref()
Out[32]: 
<output omitted>

In [33]: var('x1 x2 x3 x4 x5 x6 x6 x7 x8')
Out[33]: (x₁, x₂, x₃, x₄, x₅, x₆, x₆, x₇, x₈)

In [34]: solve_linear_system(e, x1, x2, x3, x4, x5, x6, x7, x8)
Out[34]: 
⎧    827      1377      886      1546      1171      1831      1967      2517⎫
⎨x₁: ───, x₂: ────, x₃: ───, x₄: ────, x₅: ────, x₆: ────, x₇: ────, x₈: 
────⎬
⎩    209      209       209      209       209       209       209       209 ⎭

In [35]: %timeit e.rref()
10 loops, best of 3: 65 ms per loop

In [36]: %timeit solve_linear_system(e, x1, x2, x3, x4, x5, x6, x7, x8)
10 loops, best of 3: 84.5 ms per loop
```

Original issue for #4949: http://code.google.com/p/sympy/issues/detail?id=1850
Original author: https://code.google.com/u/asmeurer@gmail.com/
Original owner: https://code.google.com/u/asmeurer@gmail.com/

**Comment by oscarbenjamin** [comment]
Using `rref` in `solve_linear_system` would also fix #6420 and #18531

**Comment by oscarbenjamin** [comment]
This can probably also fix #6592

**Comment by oscarbenjamin** [comment]
Related #10000 and #10001
